rootfs/etc/nginx/lua/monitor.lua

```diff
@@ -27,12 +27,43 @@ local function send(payload)
   assert(s:close())
 end
 
+-- when the traffic was send to canary service, the canary service name is more
+-- appropriate for metrics.
+local function get_service_name(namespace, serviceName, alternativeUpstreamName)
+  local svcName = serviceName
+  -- the variable indicates the traffic if be send to the canary service
+  if namespace and alternativeUpstreamName and #alternativeUpstreamName > #namespace then
+    -- the format of alternativeUpstreamName is `ns-svc-port`
+    local endPos = #alternativeUpstreamName
+    while endPos > 1 do
+      if string.sub(alternativeUpstreamName, endPos, endPos) == "-" then
+        endPos = endPos - 1
+        break
+      end
+      endPos = endPos - 1
+    end
```

Comment:
Why not use `ngx.re.match` instead?

Reply:
The regular expression of serviceName is `[a-z]([-a-z0-9]*[a-z0-9])?`  which is difficult to distinguish from `alternativeUpstreamName ` .